lib/FrontendTool/ScanDependencies.cpp

```diff
@@ -58,13 +59,51 @@ static void findAllImportedClangModules(ASTContext &ctx, StringRef moduleName,
   }
 }
 
+struct InterfaceSubASTContextDelegate: SubASTContextDelegate {
+  bool runInSubContext(ASTContext &ctx, StringRef interfacePath,
```

Comment:
Much of this logic is also in `ModuleInterfaceBuilder::buildSwiftModuleInternal`, although some bits of validation there are missing here. Is there a way to unify this with `buildSwiftModuleInternal`? For example, by parameterizing it with an `action` like this function has, and having the dependency scanner's action be "walk the imports" and the module builder's action be "type check and serialialize"? More sharing there would help us avoid any discrepancies with the main build.

Reply:
Yeah, it seems to be feasible. The configuration setup here is mostly for parsing to be performed correctly, which doesn't require options like `SerializationOptions` to be configured. We could potentially move these additional setups to the action body of "type check and serialize".

Reply:
I'd certainly prefer that. Anything that differs between the fast dependency scanner and the normal build process is going to be a hard-to-diagnose bug later on down the line.